File: source/javascript/bracket_matching.js
function isBalanced(s) {
    const stack = [];
    for (const c of s) {
        if (c === '(' || c === '{' || c === '[') stack.push(c);
        else if (c === ')' || c === '}' || c === ']') {
            if (stack.length === 0) return false;
            const top = stack.pop();
            if (c === ')' && top !== '(') return false;
            if (c === '}' && top !== '{') return false;
            if (c === ']' && top !== '[') return false;
        }
    }
    return stack.length === 0;
}

const tests = ["({[]})", "({[}])", ""];
for (const t of tests) {
    console.log(`"${t}" -> ${isBalanced(t) ? "balanced" : "not balanced"}`);
}

File: source/python/bracket_matching.py
def is_balanced(s):
    stack = []
    for c in s:
        if c in "({[":
            stack.append(c)
        elif c in ")}]":
            if not stack:
                return False
            top = stack.pop()
            if c == ")" and top != "(":
                return False
            if c == "}" and top != "{":
                return False
            if c == "]" and top != "[":
                return False
    return len(stack) == 0

tests = ["({[]})", "({[}])", ""]
for t in tests:
    status = "balanced" if is_balanced(t) else "not balanced"
    print(f'"{t}" -> {status}')
